src/scenechange/mod.rs

```diff
@@ -301,24 +295,29 @@ impl<T: Pixel> SceneChangeDetector<T> {
   ) -> ScenecutResult {
     // Downscaling both frames for comparison
     // Moving scaled frames to buffer
-    if self.frame_buffer.is_empty() {
-      let frame1_scaled = frame1.planes[0].downscale(self.scale_factor);
-      self.frame_buffer.push(frame1_scaled);
-
-      let frame2_scaled = frame2.planes[0].downscale(self.scale_factor);
-      self.frame_buffer.push(frame2_scaled);
+    if let Some(frame_buffer) = self.frame_buffer.as_deref_mut() {
+      frame_buffer.swap(0, 1);
+      frame2.planes[0]
+        .downscale_in_place(self.scale_factor, &mut frame_buffer[1]);
     } else {
-      self.frame_buffer.remove(0);
-      self.frame_buffer.push(frame2.planes[0].downscale(self.scale_factor));
+      self.frame_buffer = Some(Box::new([
+        frame1.planes[0].downscale(self.scale_factor),
+        frame2.planes[0].downscale(self.scale_factor),
+      ]));
```

Comment:
I'm a little unsure what you mean -- what would it look like for the frame buffer to be extracted from the conditions? Do you mean for the `Box` to be reused even after it is cleared (which is what the original code does, since `.clear()`ing a `Vec` removes the elements, but keeps the capacity which will avoid future allocations caused by this branch)?

Or do you mean for `delta_in_planes` to be inside both branches, so as to avoid using `Option::unwrap` by being able to directly reference the original frame buffer?

Reply:
Something like
```
    let frame_buffer = if let Some(frame_buffer) = self.frame_buffer.as_deref_mut() {
      frame_buffer.swap(0, 1);
      frame2.planes[0]
         .downscale_in_place(self.scale_factor, &mut frame_buffer[1]);
      &frame_buffer
    } else {
      self.frame_buffer.remove(0);
      self.frame_buffer.push(frame2.planes[0].downscale(self.scale_factor));
      let frame_buffer = Box::new([
        frame1.planes[0].downscale(self.scale_factor),
        frame2.planes[0].downscale(self.scale_factor),
      ]);
      self.frame_buffer = Some(frame_buffer);
      &frame_buffer
    }
```

Reply:
That's a great idea that I didn't think of, but unfortunately it doesn't seem to work with the borrow checker. Doing this:
```rust
    let frame_buffer =
      if let Some(frame_buffer) = self.frame_buffer.as_deref_mut() {
        frame_buffer.swap(0, 1);
        frame2.planes[0]
          .downscale_in_place(self.scale_factor, &mut frame_buffer[1]);
        todo!()
      } else {
        let frame_buffer = Box::new([
          frame1.planes[0].downscale(self.scale_factor),
          frame2.planes[0].downscale(self.scale_factor),
        ]);
        self.frame_buffer = Some(frame_buffer);
        &frame_buffer
      };
```
causes the error:
```rust
error[E0382]: borrow of moved value: `frame_buffer`
   --> src/scenechange/mod.rs:310:9
    |
305 |         let frame_buffer = Box::new([
    |             ------------ move occurs because `frame_buffer` has type `Box<[v_frame::plane::Plane<T>; 2]>`, which does not implement the `Copy` trait
...
309 |         self.frame_buffer = Some(frame_buffer);
    |                                  ------------ value moved here
310 |         &frame_buffer
    |         ^^^^^^^^^^^^^ value borrowed here after move
```
The only way I could get it to work was by returning `self.frame_buffer.as_deref().unwrap()`, but that seems to defeat the point of returning a reference to the frame buffer within each branch, unless there is some other way to convince the compiler without `unsafe` code that `self.frame_buffer` within the branch is always `Some`.

Reply:
Even if I solved the borrow checker-related errors for the branch that heap-allocates a new frame buffer by doing `delta_in_planes` inside the branch and using an early return, the other branch still gave similar (to the original posted) borrow checker errors. Even though I am essentially certain that the `std::hint::unreachable_unchecked` cannot actually be reached, I have replaced it with its safe counterpart `unreachable!` so as to avoid `unsafe` code, since the branch likely does not affect the performance much.